state/protocol/protocol_state/epochs/statemachine.go

```diff
@@ -0,0 +1,303 @@
+			parentState.GetEpochStateID(), parentEpochState.ID())
+	}
+
+	var (
+		stateMachine StateMachine
+	)
+	candidateAttemptsInvalidEpochTransition := epochFallbackTriggeredByIncorporatingCandidate(view, params, parentEpochState)
+	if parentEpochState.InvalidEpochTransitionAttempted || candidateAttemptsInvalidEpochTransition {
+		// Case 1: InvalidEpochTransitionAttempted is true, indicating that we have encountered an invalid
+		//         epoch service event or an invalid state transition previously in this fork.
+		// Case 2: Incorporating the candidate block is itself an invalid epoch transition.
+		//
+		// In either case, Epoch Fallback Mode [EFM] has been tentatively triggered on this fork,
+		// and we must use only the `epochFallbackStateMachine` along this fork.
+		//
+		// TODO for 'leaving Epoch Fallback via special service event': this might need to change.
+		stateMachine, err = epochFallbackStateMachineFactory(view, parentEpochState)
+	} else {
+		stateMachine, err = happyPathStateMachineFactory(view, parentEpochState)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not initialize protocol state machine: %w", err)
+	}
+
+	return &EpochStateMachine{
+		activeStateMachine: stateMachine,
+		parentState:        parentState,
+		mutator:            mutator,
+		epochFallbackStateMachineFactory: func() (StateMachine, error) {
+			return epochFallbackStateMachineFactory(view, parentEpochState)
+		},
+		setups:          setups,
+		commits:         commits,
+		protocolStateDB: protocolStateDB,
+	}, nil
+}
+
+// Build schedules updates to the protocol state by obtaining the updated state from the active state machine,
+// preparing deferred DB updates and committing updated sub-state ID to the KV store.
+// ATTENTION: In mature implementation all parts of the Dynamic Protocol State will rely on the Key-Value Store as storage
+// but to avoid a large refactoring we are using a hybrid approach where only the epoch state ID is stored in the KV Store
+// but the actual epoch state is stored separately, nevertheless, the epoch state ID is used to sanity check if the
+// epoch state is consistent with the KV Store. Using this approach, we commit the epoch sub-state to the KV Store which in
+// affects the Dynamic Protocol State ID which is essentially hash of the KV Store.
+func (e *EpochStateMachine) Build() protocol.DeferredDBUpdates {
+	updatedEpochState, updatedStateID, hasChanges := e.activeStateMachine.Build()
+	dbUpdates := e.pendingDbUpdates
+	dbUpdates.Add(func(tx *transaction.Tx, blockID flow.Identifier) error {
+		return e.protocolStateDB.Index(blockID, updatedStateID)(tx)
+	})
+	if hasChanges {
+		dbUpdates.AddBadgerUpdate(operation.SkipDuplicatesTx(e.protocolStateDB.StoreTx(updatedStateID, updatedEpochState)))
+	}
+	e.mutator.SetEpochStateID(updatedStateID)
+
+	return dbUpdates
+}
+
+// ProcessUpdate applies the state changes that are delivered via sealed service events.
```

Comment:
> this function must be invoked even if there are no "updates" to pass in

Thanks Jordan. I think this is a really important concern. My preference would be to mediate this risk through a callout in the goDoc and maybe renaming the function. I think that would be sufficient.

Reply:
suggestion

```golang
// EvolveState evolves the Epoch state in accordance with the information in the candidate
// block (under construction). Information that potentially changes the Epoch state is:
//   - Service Events sealed in the candidate block
//   - the candidate block's view (already provided at construction time)
//
// CAUTION: This function MUST be called for all candidate blocks, even if `update` is empty! This
// is because also the absence of expected service events by a certain view can result in the Epoch
// state changing (for example not having received the EpochCommit event for the next epoch, but
// approaching the end of the current epoch).
//
// The block's payload might contain epoch preparation service events for the next
// epoch. In this case, we need to update the tentative protocol state.
// We need to validate whether all information is available in the protocol
// state to go to the next epoch when needed. In cases where there is a bug
// in the smart contract, it could be that this happens too late, and we should trigger epoch fallback mode.
// No errors are expected during normal operations.
func (e *EpochStateMachine) EvolveState(sealedServiceEvents []flow.ServiceEvent) error {
```

besides the extended goDoc, I am suggesting the following refinements:
* rename the input `update` to `sealedServiceEvents` for clarity 
* suggesting to rename the function to `EvolveState`. I somewhat feel that this naming provides a clearer separation between the inputs and what the method does (previously we used "update" both in the function name as well as the input, which I think is part of the concern here).

Reply:
_Change of opinion_ 😅 

I think the patter of "this function must be called even if there is no input" is a vastly unsafe pattern, because it creates this additional edge case across multiple abstraction layers. I think we would need `CAUTION` statements at [this layer ](https://github.com/onflow/flow-go/blob/63d0114b1335b16d561854dfd027fc6d66bc8d2b/state/protocol/protocol_state.go#L180)for example. I think we can simplify the usage patterns of our API in this regards. I another words, I think we need to fix it (`TODO` - not in this PR)

Reply:
I think we are getting way down to a rabbit hole with this reasoning. We are as engineers are responsible for calling functions in correct order with correct arguments. To me there is little value is adding extra constructions to ensure that `EvolveState` will be called, there is exactly one place where it is called, it's part of encapsulated logic, it's not accessible to client or user in any way. To me it seems that we are trying to invent a mechanism that will enforce that `EvolveState` is called for each state machine, but that is unnecessary since the calling code is very straight-forward, it's literally a for loop which is initiated as part of the pipeline that _we_ have built. Happy to continue this discussion.